### ISSUE TITLE
ChainSync client cannot rollback to an EBB that shares a slot with an actual block

### ISSUE THREAD

**Issue by nfrisby** [input-output-hk/ouroboros-network#1352]
In this code

https://github.com/input-output-hk/ouroboros-network/blob/e9453b316dd47d54ec640cac312d6bf79957b31d/ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs#L332-L334

Suppose `intersection :: Point blk` is an EBB and there's also a proper block after it in that same slot.

While the `AF.rollback` can use the hash to distinguish the two blocks, the `rewindChainState` only has the slot number and not the hash and so will incorrectly retain the block after the EBB even though that block is after the intersection point.

**Comment by mrBliss** [comment]
We have decided on the following approach: change the `SlotNo` argument passed to `rewindChainState` to a `Point` (more precisely, a `CanValidate p hdr => Point hdr`). Quoting @nfrisby:
> We could instead change its `SlotNo` argument to a `Point` and have the chain state retain hashes. That'd also be viable and wouldn't explicitly involve EBBs (it'd only explicitly allow for multiple blocks in a slot, which we (and the PBFT code) know currently only happens with EBBs.)

And myself:
> Concretely, this means that `PBFT.ChainState` will track the hash of each applied EBB, so we can intercept a `rewindChainState` call to an EBB. The hash is ignored for regular blocks. Hmm, it feels a bit wrong to pass in the hash if we're not using it to check whether it matches a previously applied header. Not that I'm arguing we should, as that would mean that we'd have to store the hashes too, requiring even more bytes, and breaking binary compatibility

**Comment by edsko** [comment]
Thomas will have the details, but perhaps a better approach here is to follow what we agreed before in https://github.com/input-output-hk/ouroboros-network/issues/925#issuecomment-527184811 and just disallow rolling back to EBBs altogether.

**Comment by mrBliss** [comment]
> Thomas will have the details, but perhaps a better approach here is to follow what we agreed before in [#925 (comment)](https://github.com/input-output-hk/ouroboros-network/issues/925#issuecomment-527184811) and just disallow rolling back to EBBs altogether.

The details:
* Revert all the changes to `PBFT.ChainState` after #1307.
* Pass a `isEBB :: Header blk -> IsEBB` to the `ChainSyncClient` that is used to avoid rolling back to an EBB.
  + We never have to rewind our `ChainState` to an EBB.
  + Rolling back to an EBB is disallowed.
  + `findIntersection` should not send a `Point` corresponding to an EBB.
  + `intersectFound` should disconnect with `InvalidIntersection` if the intersection is an EBB.
  + `rollBackward` should disconnect with `InvalidRollBack` if the intersection is an EBB.
  + To be able to check whether a point corresponds to an EBB, the anchor point of the candidate fragment can never be an EBB. This is because we won't have the header of the anchor point in memory and thus can't check whether it's an EBB with `isEBB :: Header blk -> IsEBB`. Take care of this when trimming the candidate fragment in `intersectsWithCurrentChain`.

**Comment by nfrisby** [comment]
I think the following scenario explains how a rollback to an EBB might arise in the deployed net. (They arise more simply in the tests because our current tests are actually forging new EBBs, which won't happen in deployment.) This concrete motivator might inform details of the related special-cases in the ChainSync client.

- Suppose a node N1 somehow gets truncated back to a historical EBB.
- Suppose also that it immediately leads. So it's now on a one-block fork that intersects with the main chain at that EBB.
- At some point, it'll choose to switch back to the main chain, since that's hundreds of blocks ahead.
- To do so, it will switch chains, and the intersection of its old and new chain is the EBB.

However, since it was isolated (no other node is interested in blocks that old), that scenario doesn't actually involve any rollbacks/rewinds. (Right?)

Now suppose instead that N1 and N2 both somehow get simultaneously truncated back to the same EBB, N1 leads, and N2 follows. In this case, their respective ChainSync mini protocol instances will need to roll back & rewind to that EBB when they both eventually switch over to the main chain again.

If their CS clients crash in that circumstance, it's fine: the error policy will restart those mini protocol instances. But if they misbehave without crashing, then the N1 and N2 ChainSync mini protocol instances might end up in an incorrect/out-of-sync state. Thus we must ensure ChainSync handles rollbacks to EBBs correctly (e.g. by carefully avoiding them).

**Comment by nfrisby** [comment]
> To be able to check whether a point corresponds to an EBB, the anchor point of the candidate fragment can never be an EBB. This is because we won't have the header of the anchor point in memory and thus can't check whether it's an EBB with `isEBB :: Header blk -> IsEBB`. Take care of this when trimming the candidate fragment in `intersectsWithCurrentChain`.

I'm not seeing how to :(. In `intersectsWithCurrentChain`, we have `ourFrag' <- getCurrentChain`, so the anchor point is still just a point. And this function is in `STM m`, so -- unless I'm missing something -- we'd have to expand the ChainDB API in order to fetch that point in order to check if it's an EBB.

Please advise.

**Comment by mrBliss** [comment]
The new plan (actually the previous plan): https://github.com/input-output-hk/ouroboros-network/issues/1352#issuecomment-566433832

@nfrisby Sorry for sending you down the wrong path :grin:

**Comment by edsko** [comment]
> @nfrisby Sorry for sending you down the wrong path 😁

My fault 😳

**Comment by nfrisby** [comment]
Edsko on Slack:

> so our grudging conclusion is that modifying the PBftChainState is the right thing to do after all. Having rewindChainState take a point sounds like a reasonable thing to do, not EBB specific, and Thomas says that the changes to the codebase to do that are actually relatively minor. I’m not terribly happy about this, we got PBftFChainState wrong a number of times and it’s security sensitive, but I think it’s the best solution
> did I mention I don’t like EBBs?

Summary: we're back to the solution that was implemented in recent PRs to PBFT ChainState and culminated in PR 1353.
